internal/backup_mark.go

```diff
@@ -49,6 +103,101 @@ func GetImpermanentBackupMetadataBefore(baseBackupFolder storage.Folder, backupN
 	return previousImpermanentBackupMetadata, nil
 }
 
+func getMarkedImpermanentBackupMetadata(folder storage.Folder, backupName string) ([]UploadObject, error){
+	backupMetadata := []UploadObject{}
+	baseBackupFolder := folder.GetSubFolder(utility.BaseBackupPath)
+
+	// retrieve current backup meta
+	backup := NewBackup(baseBackupFolder, backupName)
+
+	meta, err := backup.FetchMeta()
+	if err != nil {
+		return nil, err
+	}
+
+	permanentBackups, _ := getPermanentObjects(folder)
+	//  del current backup from
+	delete(permanentBackups, backupName[len(utility.BackupNamePrefix):len(utility.BackupNamePrefix)+24])
+
+	reverseLinks, err := getGraphFromBaseToIncrement(folder)
+	if err != nil {
+		return nil, err
+	}
+
+	if backupHasPermanentInFuture(&reverseLinks, backupName, &permanentBackups){
+		return nil, NewBackupHasPermanentBackupInFutureError(backupName)
+	}
+
+	meta.IsPermanent = false
+	metadataUploadObject, err := getMetadataUploadObject(backup.Name, meta)
+	if err != nil {
+		return nil, err
+	}
+	backupMetadata = append(backupMetadata, metadataUploadObject)
+
+	return backupMetadata, nil
+
+}
+
+func backupHasPermanentInFuture(reverseLinks *map[string][]string, backupName string, permanentBackups *map[string]bool) (bool) {
+	if _, ok := (*permanentBackups)[backupName[len(utility.BackupNamePrefix):len(utility.BackupNamePrefix)+24]]; ok {
+		return true
+	}
+
+	if _, ok := (*reverseLinks)[backupName]; !ok {
```

Comment:
It appears to be the only needed check. Because if there is permanent backup somewhere in future, then one of next backups is permanent. So you can remove other if's in this function, but leave the comment.